org-code-javabuilder/theater/src/main/java/org/code/theater/GifWriter.java

```diff
@@ -0,0 +1,113 @@
+package org.code.theater;
+
+import java.awt.image.BufferedImage;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import javax.imageio.*;
+import javax.imageio.metadata.IIOInvalidTreeException;
+import javax.imageio.metadata.IIOMetadata;
+import javax.imageio.metadata.IIOMetadataNode;
+import javax.imageio.stream.ImageOutputStream;
+import org.code.protocol.InternalErrorKey;
+import org.code.protocol.InternalJavabuilderError;
+
+/**
+ * Writer to generate a gif from a set of images. The gif will be stored in the given
+ * ByteArrayOutputStream.
+ */
+public class GifWriter {
+  private ImageWriter writer;
+  private ImageWriteParam params;
+  private ImageOutputStream imageOutputStream;
+
+  public GifWriter(ByteArrayOutputStream out) {
+    try {
+      this.imageOutputStream = ImageIO.createImageOutputStream(out);
+      this.writer = ImageIO.getImageWritersBySuffix("gif").next();
+      this.params = writer.getDefaultWriteParam();
+      this.writer.setOutput(this.imageOutputStream);
+      this.writer.prepareWriteSequence(null);
+    } catch (IOException e) {
+      throw new InternalJavabuilderError(InternalErrorKey.INTERNAL_RUNTIME_EXCEPTION, e.getCause());
+    }
+  }
+
+  /**
+   * Write the given image as the next frame of the gif.
+   *
+   * @param img BufferedImage
+   * @param delay delay in milliseconds until the next frame of the gif.
```

Comment:
Does the delay come before or after the current frame that is being added?

Reply:
it's after the frame being added

Reply:
I'll make the comment more clear